go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -0,0 +1,484 @@
+		// Obtain the VIPs associated to the Service from the Service Tracker
+		vips := c.serviceTracker.getService(name, namespace)
+		for vipKey := range vips {
+			// the VIP is stored with the format IP:Port/Protocol
+			vip, proto := splitVirtualIPKey(vipKey)
+			// Get the OVN LoadBalancer
+			lbID, err := loadbalancer.GetOVNKubeLoadBalancer(proto)
+			if err != nil {
+				klog.Errorf("Error getting OVN LoadBalancer for protocol %s", proto)
+				return err
+			}
+			// Delete the Service VIP from OVN
+			klog.Infof("Deleting service %s on namespace %s from OVN", name, namespace)
+			if err := loadbalancer.DeleteLoadBalancerVIP(lbID, vip); err != nil {
+				klog.Errorf("Error deleting VIP %s on OVN LoadBalancer %s", vip, lbID)
+				return err
+			}
+			// Delete the Service VIP from the Service Tracker
+			c.serviceTracker.deleteServiceVIP(name, namespace, vip, proto)
+		}
+		// Delete the Service form the Service Tracker
+		c.serviceTracker.deleteService(name, namespace)
+		return nil
+	}
+	klog.Infof("Creating service %s on namespace %s on OVN", name, namespace)
+	// The Service exists in the cache: update it in OVN
+	// Get the endpoint slices associated to the Service
+	esLabelSelector := labels.Set(map[string]string{
+		discovery.LabelServiceName: name,
+	}).AsSelectorPreValidated()
+	endpointSlices, err := c.endpointSliceLister.EndpointSlices(namespace).List(esLabelSelector)
+	if err != nil {
+		// Since we're getting stuff from a local cache, it is basically impossible to get this error.
+		c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToListEndpointSlices",
+			"Error listing Endpoint Slices for Service %s/%s: %v", namespace, name, err)
+		return err
+	}
+	for _, ip := range service.Spec.ClusterIPs {
+		for _, svcPort := range service.Spec.Ports {
+			// ClusterIP
+			lbID, err := loadbalancer.GetOVNKubeLoadBalancer(svcPort.Protocol)
+			if err != nil {
+				c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToGetOVNLoadBalancer",
+					"Error trying to obtain the OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
+				return err
+			}
+			// create the vip
+			vip := util.JoinHostPortInt32(ip, svcPort.Port)
+			klog.V(4).Infof("Updating service %s/%s with VIP %s %s", name, namespace, vip, svcPort.Protocol)
+			// get the endpoints associated to the vip
+			family := v1.IPv4Protocol
+			if utilnet.IsIPv6String(ip) {
+				family = v1.IPv6Protocol
+			}
+			eps := getLbEndpoints(endpointSlices, svcPort, family)
+			// Reconcile OVN
+			if err := loadbalancer.UpdateLoadBalancer(lbID, vip, eps); err != nil {
+				c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
+					"Error trying to update OVN LoadBalancer for Service %s/%s: %v", name, namespace, err)
+				return err
+			}
+			c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol)
+			// TODO: Insert a reject ACL if the service doesn't have endpoints associated
+			if len(eps) == 0 {
```

Comment:
Shouldn't this check be performed before L267? I.e before doing `loadbalancer.UpdateLoadBalancer(lbID, vip, eps)` which might error if `eps` is empty?

Reply:
there are vips without endpoints, that is legit, the only requirement is that those VIPs must reject the connections ... working in implementing this TODO now